### PR TITLE
Update ikea_e1812.yaml

### PR DESCRIPTION
**Pull request by vandalon**
So, here we go....

Thinks I tried to accomplish: 
- Full de-bouncing, no restart because of double events
- Added an option that the initial press on a double press reacts instantly. (example use case; first press gives instant 50% brightness, 2nd (double press) gives 100% brightness)
- Tried to make everything according to automation standards. (Exmaple. i moved choose items with only defaults to choose items with an "empty" condition). 

Let me know what you think.

### CODE REVIEW

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 1]
```diff
@@ -1,40 +1,51 @@
 # Blueprint metadata
 blueprint:
   name: Controller - IKEA E1812 Shortcut button
-  description: |
-    Controller automation for executing any kind of action triggered by an IKEA E1812 Shortcut button. Allows to optionally loop an action on a button long press.
+  description: 'Controller automation for executing any kind of action triggered by
+    an IKEA E1812 Shortcut button. Allows to optionally loop an action on a button
+    long press.
+
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:7]:
```suggestion
  description: |
    Controller automation for executing any kind of action triggered by an IKEA E1812 Shortcut button. Allows to optionally loop an action on a button long press.
```

Keeping the current format here

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 2]
```diff
@@ -1,40 +1,51 @@
...
 
-    Automations created with this blueprint can be connected with one or more [Hooks](https://epmatt.github.io/awesome-ha-blueprints/docs/blueprints/hooks) supported by this controller.
-    Hooks allow to easily create controller-based automations for interacting with media players, lights, covers and more.
+
+    Automations created with this blueprint can be connected with one or more [Hooks](https://epmatt.github.io/awesome-ha-blueprints/docs/blueprints/hooks)
+    supported by this controller.
+
+    Hooks allow to easily create controller-based automations for interacting with
+    media players, lights, covers and more.
+
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:16]:
```suggestion
    Automations created with this blueprint can be connected with one or more [Hooks](https://epmatt.github.io/awesome-ha-blueprints/docs/blueprints/hooks) supported by this controller.
    Hooks allow to easily create controller-based automations for interacting with media players, lights, covers and more.
```

Keeping the current format here

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 3]
```diff
@@ -1,40 +1,51 @@
...
 
     Full documentation for this controller and list of available Hooks [here](https://epmatt.github.io/awesome-ha-blueprints/docs/blueprints/controllers/ikea_e1812).
-  source_url: https://github.com/EPMatt/awesome-ha-blueprints/blob/main/blueprints/controllers/ikea_e1812/ikea_e1812.yaml
+
+    '
+  source_url: https://github.com/EPMatt/awesome-ha-blueprints/blob/issue-22/blueprints/controllers/ikea_e1812/ikea_e1812.yaml
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:21]:
```suggestion
source_url: https://github.com/EPMatt/awesome-ha-blueprints/blob/main/blueprints/controllers/ikea_e1812/ikea_e1812.yaml
```

Keeping the current format here, and using the `main` branch for the `source_url`

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 4]
```diff
@@ -1,40 +1,51 @@
...
   domain: automation
   input:
     integration:
       name: Integration
-      description: Integration used for connecting the remote with Home Assistant. Select one of the available values.
+      description: Integration used for connecting the remote with Home Assistant.
+        Select one of the available values.
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:27]:
```suggestion
      description: Integration used for connecting the remote with Home Assistant. Select one of the available values.
```

Keeping the current format here

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 5]
```diff
@@ -1,40 +1,51 @@
...
       selector:
         select:
           options:
-            - deCONZ
-            - ZHA
-            - Zigbee2MQTT
+          - deCONZ
+          - ZHA
+          - Zigbee2MQTT
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:33]:
```suggestion
            - deCONZ
            - ZHA
            - Zigbee2MQTT
```

Keeping the current format here

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 6]
```diff
@@ -1,40 +1,51 @@
...
     controller_device:
       name: (deCONZ, ZHA) Controller Device
-      description: The controller device to use for the automation. Choose a value only if the remote is integrated with deCONZ, ZHA.
+      description: The controller device to use for the automation. Choose a value
+        only if the remote is integrated with deCONZ, ZHA.
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:37]:
```suggestion
      description: The controller device to use for the automation. Choose a value only if the remote is integrated with deCONZ, ZHA.
```

Keeping the current format here

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 7]
```diff
@@ -1,40 +1,51 @@
...
       default: ''
       selector:
-        device:
+        device: {}
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:40]:
```suggestion
        device:
```
The empty object notation is not required here

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 8]
```diff
@@ -1,40 +1,51 @@
...
     controller_entity:
       name: (Zigbee2MQTT) Controller Entity
-      description: The action sensor of the controller to use for the automation. Choose a value only if the remote is integrated with Zigbee2MQTT.
+      description: The action sensor of the controller to use for the automation.
+        Choose a value only if the remote is integrated with Zigbee2MQTT.
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:44]:
```suggestion
      description: The action sensor of the controller to use for the automation. Choose a value only if the remote is integrated with Zigbee2MQTT.
```

Keeping the current format here

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 9]
```diff
@@ -1,40 +1,51 @@
...
       default: ''
       selector:
         entity:
           domain: sensor
-    # inputs for custom actions
     action_button_short:
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:49]:
```suggestion
    # inputs for custom actions
    action_button_short:
```

keep comment since it improves code readability

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 10]
```diff
@@ -59,197 +70,279 @@ blueprint:
       default: []
       selector:
         action:
-    # inputs for looping custom actions on long button press events until the corresponding release event is received
     button_long_loop:
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:73]:
```suggestion
    # inputs for looping custom actions on long button press events until the corresponding release event is received
    button_long_loop:
```

keep comment since it improves code readability

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 11]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
     deconz:
-      button_short: [1002]
-      button_long: [1001]
-      button_release: [1003]
+      button_short:
+      - 1002
+      button_long:
+      - 1001
+      button_release:
+      - 1003
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:162]:
```suggestion
      button_short: [1002]
      button_long: [1001]
      button_release: [1003]
```

keep the inline format since it improves readability

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 12]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
     zha:
-      button_short: ['on']
-      button_long: [move_with_on_off]
-      button_release: [stop]
+      button_short:
+      - 'on'
+      button_long:
+      - move_with_on_off
+      button_release:
+      - stop
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:169]:
```suggestion
      button_short: ['on']
      button_long: [move_with_on_off]
      button_release: [stop]
```

keep the inline format since it improves readability

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 13]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
-      button_long: [brightness_move_up]
-      button_release: [brightness_stop]
-  # pre-choose actions for buttons based on configured integration
-  # no need to perform this task at automation runtime
+      button_short:
+      - 'on'
+      button_long:
+      - brightness_move_up
+      button_release:
+      - brightness_stop
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:176]:
```suggestion
      button_short:
      - 'on'
      button_long:
      - brightness_move_up
      button_release:
      - brightness_stop
  # pre-choose actions for buttons based on configured integration
  # no need to perform this task at automation runtime
```

keep the inline format since it improves readability, bring back comments

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 14]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
-  controller_device: !input controller_device
-  controller_id: '{% if integration_id=="zigbee2mqtt" %}{{controller_entity}}{% else %}{{controller_device}}{% endif %}'
-# mode restart ensures hold actions are interrupted when the corresponding release action is received
+  integrations_with_debounce:
+  - deconz
+  - zha
+  controller_entity: !input 'controller_entity'
+  controller_device: !input 'controller_device'
+  controller_id: '{% if integration_id=="zigbee2mqtt" %}{{controller_entity}}{% else
+    %}{{controller_device}}{% endif %}'
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:186]:
```suggestion
  integrations_with_debounce: [deconz, zha]
  # build data to send within a controller event
  controller_entity: !input 'controller_entity'
  controller_device: !input 'controller_device'
  controller_id: '{% if integration_id=="zigbee2mqtt" %}{{controller_entity}}{% else %}{{controller_device}}{% endif %}'
  # mode restart ensures hold actions are interrupted when the corresponding release action is received
```

use inline array format, bring back comments

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 15]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
+- platform: event
+  event_type: state_changed
+  event_data:
+    entity_id: !input 'controller_entity'
+- platform: event
+  event_type:
+  - zha_event
+  - deconz_event
+  event_data:
+    device_id: !input 'controller_device'
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:199]:
```suggestion
# trigger for zigbee2mqtt
- platform: event
  event_type: state_changed
  event_data:
    entity_id: !input 'controller_entity'
# trigger for other integrations
- platform: event
  event_type:
  - zha_event
  - deconz_event
  event_data:
    device_id: !input 'controller_device'
```

bring back comments

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 16]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
+        boolean:
+    button_double_initial_press:
+      name: Run short press action at once.
+      description: Choose whether or not to run the short press action at once, without waiting for the 2nd press.
+        When a second press is received the double press action will be run at once and the short press action will stop if it's still running.
       default: false
       selector:
         boolean:
-    # helpers used to properly recognize the remote button events
+    #Only requiered since it seems impossible to get the automations attribute dynamicly via a blueprint.
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:104]:
```suggestion
    # helpers used to properly recognize the remote button events
```

we can keep the old comment here, so it won't confuse users. also, this way we're keeping the same format for all blueprints

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 17]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
+          min: 0
+          max: 5000
+          mode: slider
+          step: 1
     button_double_press:
       name: Expose button double press event
-      description: Choose whether or not to expose the virtual double press event for the button. Turn this on if you are providing an action for the button double press event.
+      description: Choose whether or not to expose the virtual double press event
+        for the button. Turn this on if you are providing an action for the button
+        double press event.
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:93]:
```suggestion
      description: Choose whether or not to expose the virtual double press event for the button. Turn this on if you are providing an action for the button double press event.
```

keep current format

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 18]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
+      name: Long Button press - Maximum loops
+      description: Maximum amount of repeats.
+      default: 500
+      selector:
+        number:
+          min: 0
+          max: 5000
+          mode: slider
+          step: 1
     button_double_press:
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:89]:
```suggestion
    # inputs for enabling double press events
    button_double_press:
```

add back comment

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 19]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
+        When a second press is received the double press action will be run at once and the short press action will stop if it's still running.
       default: false
       selector:
         boolean:
-    # helpers used to properly recognize the remote button events
+    #Only requiered since it seems impossible to get the automations attribute dynamicly via a blueprint. 
     helper_last_controller_event:
-      name: (Double press action Required) Helper - Last Controller Event
-      description: Input Text used to store the last event fired by the controller. See docs for more.
+      name: (Requieded to counter double triggers) Helper - Last Controller Event
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:106]: not that clear
```suggestion
      name: (Double press action Required) Helper - Last Controller Event
```

I think we should find a different sentence to put into round brackets here. "Double press action Required" is maybe the best one, since it does not expose implementation details to the blueprint user: it simply says it is required in this case.

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 20]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
+      description: Input Text used to store the last event fired by the controller.
+        See docs for more.
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:108]:
```suggestion
      description: Input Text used to store the last event fired by the controller. See docs for more.
```

keep the same format here

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml [hunk 21]
```diff
@@ -59,197 +70,279 @@ blueprint:
...
       default: ''
       selector:
         entity:
-          domain: input_text
-    helper_double_press_delay:
+          domain: input_text            
+    double_press_delay:
```
Comment (EPMatt) [inline blueprints/controllers/ikea_e1812/ikea_e1812.yaml:113]:
```suggestion
          domain: input_text
    helper_double_press_delay:
```

keep the same format. Moreover, I'd prefer not to rename current inputs, if not strictly required, since it requires users which already are using this blueprint to reconfigure their automations from scratch, since the YAML config won't be valid anymore

**Review by EPMatt — CHANGES_REQUESTED** [review]
Hi @vandalon, thank you so much for your contribution, really appreciated!

I've added a few code suggestions here, would you mind reviewing them? 
I still have to look at the second part of the blueprint, but I'd prefer to first solve this review.
I noticed code suggestions tend to overcrowd the PR, so maybe to break them in 2 reviews. 

Also my first time in using code reviews, so there's probably a better way to perform this, I should find out :) 

Thank you again for your help!

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml (hunk 1 above)
Comment (vandalon) [reply blueprints/controllers/ikea_e1812/ikea_e1812.yaml:7]: This somehow came over from the test yaml you posted earlier. I did not change this on purpose :)

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml (hunk 7 above)
Comment (vandalon) [reply blueprints/controllers/ikea_e1812/ikea_e1812.yaml:29]: This was actually on purpose. the syntax checker for HA blueprints complained about this.

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml (hunk 21 above)
Comment (vandalon) [reply blueprints/controllers/ikea_e1812/ikea_e1812.yaml:99]: A helper is a input_something in home assistant, the delay is a blueprint variable and thus not the same thing as a helper. I found this confusing :)

**Comment by vandalon** [comment]
So, basically I used your test version and worked on that, that caused more changes than needed, sorry about that. I assumed those where changes you wanted to put in anyway :)

File: blueprints/controllers/ikea_e1812/ikea_e1812.yaml (hunk 21 above)
Comment (vandalon) [reply blueprints/controllers/ikea_e1812/ikea_e1812.yaml:101]: you are right in that it would break stuff for current users. So let's keep it like that.

**Comment by vandalon** [comment]
I'm ready for the 2nd part of the review :)

**Comment by vandalon** [comment]
I just noticed that all the comments on the 1744 are also gone in my code. It seems that HA removes comments when implementing blueprints somehow?

**Comment by EPMatt** [comment]
Hi @vandalon, I'm really sorry for not reaching out to you sooner. Unfortunately I had a quite busy week and didn't have the time to look into your PR.
By the way, I'm now reviewing the second part, with the hope of merging this PR by the end of the day. :)

> I just noticed that all the comments on the 1744 are also gone in my code. It seems that HA removes comments when implementing blueprints somehow?

Yeah, HA strips comments and I think it also reformats strings using a different YAML notation, even though I'm not sure about the latter. During development, I suggest editing the blueprints in the git repository. Then for testing you can push changes to your GitHub and import the blueprint using your repo's URL, or simply copying the blueprint file into the `config/blueprints` folder or your installation.
Probably, that's why we had those unwanted changes in the first PR commit, but don't worry. :)

Again, thank you so much for your effort and time you're putting on this.

**Comment by vandalon** [comment]
Yeah, it is probably better to work "offline" on the code. I will do that for the next one, i'm already to deep into the symfonik to switch now, But will check later before submitting :P

**Comment by vandalon** [comment]
Hmmm, somehow my change/fix for bug #28 is now also in this pull request. I should learn how to use git and branches and all that I guess :).

**Comment by EPMatt** [comment]
Hi @vandalon, when working with Pull Requests, it's advised to create a separate branch just for that issue/feature you're working on.
You'll then need to create a PR from your branch against the `main` branch.  This way is your work on multiple issues in parallel you'll be able to separate commits between multiple branches and PRs. :)

You can read more about PRs and branches [here](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) and [here](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-branches).

**Review by EPMatt — APPROVED** [review]

**Comment by vandalon** [comment]
Didn't you like the initial button press option? :)

**Comment by EPMatt** [comment]
@vandalon thank you again for your contribution.

I'm opening a new PR for changes to the Light Hook (#28), restoring the previous state for the `light.yaml` here, and copying your commits in the related PR. 

The PR is currently in draft. We'll take a look at that issue later. :) 

Also I've made some changes to the `ikea_e1812` blueprint. A quick recap of what's changed:
* restored all comments and previous formatting / standards, to reduce the changes introduced with this PR only to what's required to implement the debouncing feature
* removed  `wait_for_trigger` block in the choose step for the `button_short` action, since we're currently using `mode:restart` for the automation, and waiting for a trigger there would be useful only for the `timeout` part (even if the trigger is caught, the automation would restart and subsequent actions would not run)
* restored the usage of the `helper_double_press_delay` as a delay for catching double press events
* improved the way double press events are recognized, similar to how you implemented the check for `button_long` events (very interesting! something which I didn't think about initially but which could improve the reliability of this blueprint)
* removed unused variables
* removed unused fields which were being saved in the `helper_last_controller_event`, even though they weren't used in the automation
* removed the `button_double_initial_press` feature, since I was not able to find a practical use case when users would like to have this: if a user want to double press, I think he never wants to have the `short_action` run while he's double pressing the button. Anyway that's only my opinion, if there's a valid reason to keep it we can add it back :).
* moved writing of the `helper_last_controller_event` at the top of the script, to make sure the entity gets updated asap and can be tested with a correct value by duplicate runs. 
* all the logic to implement debouncing has now been moved to the `condition` section of the automation
* implemented optimizations where possible

I hope you don't mind if I changed a few things on your PR. Of course, if what I've done does not seem right to you or doesn't solve the issue, we can discuss changes and come up with a solution together. :) 

Could you please test if the blueprint is now working correctly with your controller? 

Thakn you again for your great contribution. Really appreciate your effort 🚀

**Comment by vandalon** [comment]
In regards to the button_double_initial_press option. In some cases i like instant response and not wait for the double press timeout. For when a 2nd press would do something additional to the first press. Example:

frist press: light turns on to let's say 50% and the 2nd would set it to 100%. In this case we don't have to wait half a second to turn the light on, we can do that immediately and on the 2nd press then it goes to 100%, if received.

**Comment by vandalon** [comment]
The wait for trigger is still a good thing, even though we do mode restart. Because that will wait from that point on for any trigger without the need for a restart. and thus not going trough all the conditions of the automation again.

**Comment by vandalon** [comment]
I will test tonight btw ;)

**Comment by EPMatt** [comment]
> In regards to the button_double_initial_press option. In some cases i like instant response and not wait for the double press timeout. For when a 2nd press would do something additional to the first press. Example:
> 
> frist press: light turns on to let's say 50% and the 2nd would set it to 100%. In this case we don't have to wait half a second to turn the light on, we can do that immediately and on the 2nd press then it goes to 100%, if received.

Thank you for providing an example. This might be an interesting use case. If the double press action is disabled, the short action is executed instantly, but this does not happen when double press is enabled: we have always to get through the delay. 

In your situation the short press action is partially reaching the goal accomplished by the double press one, so having this feature available would improve system responsiveness.
Given this, we could add it back as an "advanced feature". :)

**Comment by EPMatt** [comment]
> The wait for trigger is still a good thing, even though we do mode restart. Because that will wait from that point on for any trigger without the need for a restart. and thus not going trough all the conditions of the automation again.

I'm not sure about that. Let's simulate a run for the automation when the `wait_for_trigger` block is present:

a. double press:
  1. button is pressed
  2. automation is triggered and control reaches the `wait_for_trigger` block
  3. control now waits for the second button press
  4. button is pressed for the 2nd time
  5. **the automation restarts** due to `mode: restart`. control executes the `action` sequence back from the top, and code after the `wait_for_trigger` is not executed

b. single press:
  1. button is pressed
  2. automation is triggered and control reaches the `wait_for_trigger` block
  3. control now waits for the second button press
  4. no second press is received, timeout expires. the automation proceeds to the `choose` after the `wait_for_trigger` block, resulting in the `action_button_short` being executed as expected.

Given this behaviour, the `wait_for_trigger` is not required there, since the same procedure can be implemented simply by using a `delay` block, and taking advantage of `mode: restart`.

Is this analysis correct? :)

**Comment by EPMatt** [comment]
> I will test tonight btw ;)

Thank you @vandalon, if that works as expected we can proceed to update docs and adding back the `button_double_initial_press` option. ;)

**Comment by vandalon** [comment]
> > In regards to the button_double_initial_press option. In some cases i like instant response and not wait for the double press timeout. For when a 2nd press would do something additional to the first press. Example:
> > frist press: light turns on to let's say 50% and the 2nd would set it to 100%. In this case we don't have to wait half a second to turn the light on, we can do that immediately and on the 2nd press then it goes to 100%, if received.
> 
> Thank you for providing an example. This might be an interesting use case. If the double press action is disabled, the short action is executed instantly, but this does not happen when double press is enabled: we have always to get through the delay.
> 
> In your situation the short press action is partially reaching the goal accomplished by the double press one, so having this feature available would improve system responsiveness.
> Given this, we could add it back as an "advanced feature". :)

I can always make a proper pull request for this somewhere in te future :)

**Comment by vandalon** [comment]
> > The wait for trigger is still a good thing, even though we do mode restart. Because that will wait from that point on for any trigger without the need for a restart. and thus not going trough all the conditions of the automation again.
> 
> I'm not sure about that. Let's simulate a run for the automation when the `wait_for_trigger` block is present:
> 
> a. double press:
> 
> 1. button is pressed
> 2. automation is triggered and control reaches the `wait_for_trigger` block
> 3. control now waits for the second button press
> 4. button is pressed for the 2nd time
> 5. **the automation restarts** due to `mode: restart`. control executes the `action` sequence back from the top, and code after the `wait_for_trigger` is not executed
> 
> b. single press:
> 
> 1. button is pressed
> 2. automation is triggered and control reaches the `wait_for_trigger` block
> 3. control now waits for the second button press
> 4. no second press is received, timeout expires. the automation proceeds to the `choose` after the `wait_for_trigger` block, resulting in the `action_button_short` being executed as expected.
> 
> Given this behaviour, the `wait_for_trigger` is not required there, since the same procedure can be implemented simply by using a `delay` block, and taking advantage of `mode: restart`.
> 
> Is this analysis correct? :)

The automation will not be restarted if the wait_for_trigger is triggered. I've just tested this to make sure. But if you press the button while the wait_for_trigger is active, the automation does not get triggered. 

I think it's good not to use restart if not needed :)

**Comment by vandalon** [comment]
I see you removed the debounce delay, this one is needed because otherwise any trigger will start the automation instantly and debouncing will not work. We check in the conditions if the time between triggers is bigger then the debounce time but after that now we start the automation without waiting, so if then a new trigger happens , the automation is already executed. So in before the double press we still need to build in a delay.

**Comment by vandalon** [comment]
Is it ok if I put back the wait for trigger stuff? I really think it's better :)

**Comment by vandalon** [comment]
You use the `choose:` option with only a `default`. This is not allowed by the HA syntax checker (you can try in any automation within HA). having a condition and and sequence is mandatory. This might work now, but it can break with future releases of HA because it's unsupported. So it might be a good idea to change these choices to something like:

```
- chose:
  - conditions: []
    sequence: !input whatever
```
or if you want it to be within the default:

```
- chose:
  - conditions: []
    sequence: []
  default: !input whatever
```

**Comment by vandalon** [comment]
Somehow, the time delta condition is not working as expected. I still see events coming trough where the time delta is below the threshold.

**Comment by vandalon** [comment]
Of course. because the input is set directly, without a delay in front of it, it will always run the first of multiple triggers because there the debounce delay is bigger, and we don't wait before writing to the input. So we need to wait till after the debounce delay before we fill the input field. But since we only need this for double press I can move this to the double press section of the automation.

**Comment by EPMatt** [comment]
> The automation will not be restarted if the wait_for_trigger is triggered. I've just tested this to make sure. But if you press the button while the wait_for_trigger is active, the automation does not get triggered.
> 
> I think it's good not to use restart if not needed :)

I don't think this is the expected behaviour. In my opinion, `mode:restart` should take control and restart the automation. I'll make some tests and open an Issue on HA Core, to check whether this is a bug or a feature of the `wait_for_trigger` block.

By the way, unfortunately we cannot get rid of `mode:restart`, since that allows to stop the loop for long actions before the loop is executed `button_long_max_loops` times.

**Comment by EPMatt** [comment]
> You use the `choose:` option with only a `default`. This is not allowed by the HA syntax checker (you can try in any automation within HA). having a condition and and sequence is mandatory. This might work now, but it can break with future releases of HA because it's unsupported. So it might be a good idea to change these choices to something like:
> 
> ```
> - chose:
>   - conditions: []
>     sequence: !input whatever
> ```
> 
> or if you want it to be within the default:
> 
> ```
> - chose:
>   - conditions: []
>     sequence: []
>   default: !input whatever
> ```
> 
> I have the fix almost ready in my code, so if you agree I will also apply that :)

I wasn't aware of it, thank you for pointing it out. Better to make sure we are running fully supported code, so totally agree on this, you can apply back the fix.

**Comment by vandalon** [comment]
Yes, but if we receive lets say 3 events. the first event is always going trough if you don't build in a delay. We don't want that, we want the last event and ignore everything before that if it's within the debounce_delay time. 

The first trigger writes, then another couple of triggers ar received , restarting the automation but stopping it because of the triggers beging within the trigger_delta period. We don't want the automation to be restarted, we want it to run :) So we need the act on the last trigger instead of the first.

I'm aware that we can not get rid of mode:restart, or at least i found out the hard way :) Maybe there is a way we don't know about yet though :D

**Comment by vandalon** [comment]
Actually, maybe we should move the delay to the front for everything. In some cases if a long press for example sends multiple events it could start of custom automation and then restart to do it again. It could cause strange issues in some cases

**Comment by vandalon** [comment]
I've tested this latest commit to be working properly :)

**Comment by EPMatt** [comment]
Hi @vandalon,

today I was able to get my hands on a E1812, which I successfully configured with ZHA. I'll perform a few tests with this blueprint, trying with different debouncing approaches and corner cases, checking which solution could be the best.

I'm sure your code works but I'd like to investigate if there's a simpler way to accomplish the same goal. Since this fix could then be applied to many other controllers, I'd like to make sure we're doing the right thing. My main concerns are simplicity and reliability.

Btw I've already tested my version of debouncing and noticed that, as you reported, sometimes duplicate events are raised, so something is not working as I intended. :)

I'll keep you updated here. Thank you for your time and patience, as well as the huge help you are providing to tackle this issue. ;)

**Comment by vandalon** [comment]
Hi @EPMatt ,

What do you mean with simpler? I think this method can be used for all controllers. only difference being the debounce_delay.

Simply put we always need to do the following on a trigger:

- Wait for `debounce_delay` so re-triggers restart the automation but only the last in the series executes the automation. It must be the last because otherwise an automation can be interrupted by a re-trigger. I know you don't like the initial delay :D. But this is really the only way you can counter re-triggers. Also this is more or less what debouncing means :) The thing is, with a `debounce_delay` < 200ms it will be hardly noticeable for the user. and I think in the most cases debouncing can be done way below 200ms. (I use 150ms for this switch and less then 50 for the 1744). 
- Write the trigger time to a helper so we can use that for checking how quickly a trigger is executed. (At least until there is a way to use the automation trigger time itself). 

We could leave out the condition to check for trigger_delta  because the delay picks that up anyway. But leaving it will cause all intermediate triggers within the `debounce_delay` to be ignored. Otherwise it would cause more restarts.

All the other changes are more about the double press routine. Which indeed is a lot of code for a smiple thing like that. But I don't think there is a much easier way. 

Anyway it's always fun to play with new stuff so enjoy! :)

Also I am nearly finished with adding the debouncing functionality in the E1744 code. But I can only make a new pull requests after this one because I was so stupid to use the main branch for this one :P 

Cheers!

**Comment by vandalon** [comment]
is it ok if I do this pull request again (deleting this one) so that I can do it in a specific branch? 
Because I have several other things i am working on and would like to commit them :)

**Comment by EPMatt** [comment]
Hi @vandalon,

> Hi @EPMatt ,
> 
> What do you mean with simpler? I think this method can be used for all controllers. only difference being the debounce_delay.
> 
> Simply put we always need to do the following on a trigger:
> 
> * Wait for `debounce_delay` so re-triggers restart the automation but only the last in the series executes the automation. It must be the last because otherwise an automation can be interrupted by a re-trigger. I know you don't like the initial delay :D. But this is really the only way you can counter re-triggers. Also this is more or less what debouncing means :) The thing is, with a `debounce_delay` < 200ms it will be hardly noticeable for the user. and I think in the most cases debouncing can be done way below 200ms. (I use 150ms for this switch and less then 50 for the 1744).
> * Write the trigger time to a helper so we can use that for checking how quickly a trigger is executed. (At least until there is a way to use the automation trigger time itself).
> 
> We could leave out the condition to check for trigger_delta because the delay picks that up anyway. But leaving it will cause all intermediate triggers within the `debounce_delay` to be ignored. Otherwise it would cause more restarts.
> 
> All the other changes are more about the double press routine. Which indeed is a lot of code for a smiple thing like that. But I don't think there is a much easier way.
> 
> Anyway it's always fun to play with new stuff so enjoy! :)
> 
> Also I am nearly finished with adding the debouncing functionality in the E1744 code. But I can only make a new pull requests after this one because I was so stupid to use the main branch for this one :P
> 
> Cheers!

Thank you for the detailed explanation. My only concern is having that additional delay in the sequence, which I'd like to avoid as you guessed. :D But if that's the only way to implement this feature of course I'm okey with that. :)

My idea is only to try this interesting feature on my instance and see if something can be improved (although it's currently working 100% as it is right now). Maybe it's better if we push this feature as we implemented it right now, so users can already benefit of the debouncing feature. Then we'll worry about performance and optimizations. :)

By the way, ok for opening a new PR and closing this one, so you can better organize the workflow on your fork. 👍🏻 

Thank you again for your time!
